### PR TITLE
New Resources: `azuread_application_registration`, `azuread_application_from_template` and others

### PR DESCRIPTION
**Pull request by manicminer**
### Rationale and Practitioner Benefits

We're adding these resources to complement the existing `azuread_application` resource. Whilst the existing monolithic resource is useful for declaring and managing the whole application manifest at once, these new resources offer a more composable approach and only manage a small set of properties each. They are intended for use both in simpler configurations, and in more advanced configs, the latter of which which might include:

* where application properties need to self-reference values from the application that are only known after creation (e.g. the client ID)
* where you need to create 2 or more applications that reference each other in certain property values, thereby creating a circular graph dependency
* instantiating gallery applications from template, where properties are applied to the new application but the `azuread_application` resource removes them or otherwise updates them
* where new applications are assigned an initial owner by the API but the `azuread_application` resource isn't able to detect this correctly, perhaps due to lack of API permissions
* where you might want to retain any other default property value inherited by a new application but cannot or do not wish to add these to your Terraform configuration
* where the linked service principal requires certain property values to be set before you can set a related property value on the application object
* where you want to manage individual aspect of an application across more than one Terraform module, for example assigning owners or API permissions

### Internal work

This includes some refactoring to accommodate these new resources and also to aid in moving over to new structured resource IDs

- Updating github.com/manicminer/hamilton
- Updating github.com/hashicorp/go-azure-sdk
- Updating golangci-lint
- Deprecating and removing the `utils` package which contained legacy helper funcs
- Structured ID and state migration for application resource
- Compatibility fixes for existing `azuread_application_*` resources as stopgap until v3.0
- Fixing up some validation functions

More ID parser upgrades and state migrations to follow in separate PRs.

### TODO

- [x] Upstream PR for Hamilton changes
- [x] Documentation

### Changelog

FEATURES:

* **New Resource:** `azuread_application_api_access`
* **New Resource:** `azuread_application_app_role`
* **New Resource:** `azuread_application_fallback_public_client`
* **New Resource:** `azuread_application_from_template`
* **New Resource:** `azuread_application_identifier_uri`
* **New Resource:** `azuread_application_known_clients`
* **New Resource:** `azuread_application_owner`
* **New Resource:** `azuread_application_permission_scope`
* **New Resource:** `azuread_application_redirect_uris`
* **New Resource:** `azuread_application_registration`

ENHANCEMENTS:

* `data.azuread_application` - export the `client_id` attribute, deprecate the `application_id` attribute
* `data.azuread_service_principal` - support for the `client_id` property, deprecate the `application_id` property
* `data.azuread_service_principals` - support for the `client_ids` property, deprecate the `application_ids` property
* `data.azuread_service_principals` - export the `client_id` attribute in the `service_principals` block, deprecate the `application_id` attribute
* `azuread_application` - export the `client_id` attribute, deprecate the `application_id` attribute
* `azuread_application_federated_identity_credential` - support for the `application_id` property, deprecate the `application_object_id` property
* `azuread_application_certificate` - support for the `application_id` property, deprecate the `application_object_id` property
* `azuread_application_password` - support for the `application_id` property, deprecate the `application_object_id` property
* `azuread_application_pre_authorized` - support for the `application_id` property, deprecate the `application_object_id` property
* `azuread_service_principal` - support for the `client_id` property, deprecate the `application_id` property

### Issues

Closes: #428
Closes: #501
Closes: #703
Closes: #773 
Closes: #778
Closes: #797
Closes: #800
Closes: #834
Closes: #836
Closes: #852
Closes: #919
Closes: #924 
Closes: #950
Closes: #1001
Closes: #1002
Closes: #1093

### CODE REVIEW

**Comment by manicminer** [comment]
Test results

Screenshot 2023-10-19 at 21 46 20

Screenshot 2023-10-19 at 21 37 34

File: docs/data-sources/application.md [hunk 1]
```diff
@@ -43,7 +44,8 @@ The following attributes are exported:
 * `api` - An `api` block as documented below.
 * `app_role_ids` - A mapping of app role values to app role IDs, intended to be useful when referencing app roles in other resources in your configuration.
 * `app_roles` - A collection of `app_role` blocks as documented below. For more information see [official documentation on Application Roles](https://docs.microsoft.com/en-us/azure/architecture/multitenant-identity/app-roles).
-* `application_id` - The Application ID (also called Client ID).
+* `application_id` - (Deprecated) The Client ID for the application.
```
Comment (katbyte) [inline docs/data-sources/application.md:47]: should we just remove this like we do in rm as its deprecated? / should be `deprecated in favour of X`

File: docs/data-sources/service_principal.md [hunk 2]
```diff
@@ -44,23 +44,25 @@ data "azuread_service_principal" "example" {
...
 
 ## Attributes Reference
 
 The following attributes are exported:
 
-* `account_enabled` - Whether or not the service principal account is enabled.
+* `account_enabled` - Whether the service principal account is enabled.
 * `alternative_names` - A list of alternative names, used to retrieve service principals by subscription, identify resource group and full resource ids for managed identities.
-* `application_id` - The application ID (client ID) of the application associated with this service principal.
+* `application_id` - (Deprecated) The client ID of the application associated with this service principal.
```
Comment (katbyte) [inline docs/data-sources/service_principal.md:60]: and here

File: docs/data-sources/service_principal.md [hunk 3]
```diff
@@ -44,23 +44,25 @@ data "azuread_service_principal" "example" {
...
 * `app_role_assignment_required` - Whether this service principal requires an app role assignment to a user or group before Azure AD will issue a user or access token to the application.
 * `app_role_ids` - A mapping of app role values to app role IDs, as published by the associated application, intended to be useful when referencing app roles in other resources in your configuration.
 * `app_roles` - A list of app roles published by the associated application, as documented below. For more information [official documentation](https://docs.microsoft.com/en-us/azure/architecture/multitenant-identity/app-roles).
 * `application_tenant_id` - The tenant ID where the associated application is registered.
+* `client_id` - (Deprecated) The client ID of the application associated with this service principal.
```
Comment (katbyte) [inline docs/data-sources/service_principal.md:65]: and here

File: docs/data-sources/service_principals.md [hunk 4]
```diff
@@ -76,10 +78,11 @@ The following attributes are exported:
 
 `service_principal` object exports the following:
 
-* `account_enabled` - Whether or not the service principal account is enabled.
+* `account_enabled` - Whether the service principal account is enabled.
 * `app_role_assignment_required` - Whether this service principal requires an app role assignment to a user or group before Azure AD will issue a user or access token to the application.
-* `application_id` - The application ID (client ID) of the application associated with this service principal.
+* `application_id` - (Deprecated) The client ID of the application associated with this service principal.
```
Comment (katbyte) [inline docs/data-sources/service_principals.md:83]: and here

File: docs/resources/application.md [hunk 5]
```diff
@@ -324,7 +331,8 @@ The following arguments are supported:
 In addition to all arguments above, the following attributes are exported:
 
 * `app_role_ids` - A mapping of app role values to app role IDs, intended to be useful when referencing app roles in other resources in your configuration.
-* `application_id` - The Application ID (also called Client ID).
+* `application_id` - (Deprecated) The Client ID for the application.
```
Comment (katbyte) [inline docs/resources/application.md:334]: and here

File: docs/resources/service_principal.md [hunk 6]
```diff
@@ -93,7 +93,11 @@ The following arguments are supported:
 * `account_enabled` - (Optional) Whether or not the service principal account is enabled. Defaults to `true`.
 * `alternative_names` - (Optional) A set of alternative names, used to retrieve service principals by subscription, identify resource group and full resource ids for managed identities.
 * `app_role_assignment_required` - (Optional) Whether this service principal requires an app role assignment to a user or group before Azure AD will issue a user or access token to the application. Defaults to `false`.
-* `application_id` - (Required) The application ID (client ID) of the application for which to create a service principal.
+* `application_id` - (Optional, Deprecated) The client ID of the application for which to create a service principal.
+* `client_id` - (Optional) The client ID of the application for which to create a service principal.
```
Comment (katbyte) [inline docs/resources/service_principal.md:97]: and here etc,

i'd lean towards removal, but at least should indicate client id is replacement

**Review by katbyte — APPROVED** [review]
Aside from some minor comments LGTM 🌵

**Comment by manicminer** [comment]
Hi @Israphel, unfortunately I missed these but @PseudoCoding caught them and fixed in [#1220](https://github.com/hashicorp/terraform-provider-azuread/pull/1220/files). The docs will be updated this week when the release goes out.

**Pull request merged**